src/front/glsl/ast.rs

```diff
@@ -257,64 +314,209 @@ impl Context {
+                        .expressions
+                        .append(Expression::Binary { op, left, right });
+
+                    if left_is_vector && right_is_vector {
+                        self.expressions.append(Expression::Relational {
+                            fun,
+                            argument: expr,
+                        })
+                    } else {
+                        expr
+                    }
+                } else {
+                    self.expressions
+                        .append(Expression::Binary { left, op, right })
+                }
+            }
+            ExprKind::Unary { op, expr } if !lhs => {
+                let expr = self.lower(program, *expr, false, body)?;
+
+                self.expressions.append(Expression::Unary { op, expr })
+            }
+            ExprKind::Variable(var) => {
+                if lhs {
+                    var.expr
+                } else {
+                    var.load.unwrap_or(var.expr)
+                }
+            }
+            ExprKind::Call(call) if !lhs => {
+                program.function_call(self, call.kind, call.args, body)?
+            }
+            ExprKind::Conditional {
+                condition,
+                accept,
+                reject,
+            } if !lhs => {
+                let condition = self.lower(program, *condition, false, body)?;
+                let accept = self.lower(program, *accept, false, body)?;
+                let reject = self.lower(program, *reject, false, body)?;
+
+                self.expressions.append(Expression::Select {
+                    condition,
+                    accept,
+                    reject,
+                })
+            }
+            ExprKind::Assign { tgt, value } if !lhs => {
+                let pointer = self.lower(program, *tgt, true, body)?;
+                let value = self.lower(program, *value, false, body)?;
+
+                body.push(Statement::Store { pointer, value });
+
+                value
+            }
+            _ => {
+                return Err(ErrorKind::SemanticError(
+                    format!("{:?} cannot be in the left hand side", expr).into(),
+                ))
+            }
+        })
+    }
 }
 
-#[derive(Debug)]
-pub struct ExpressionRule {
-    pub expression: Handle<Expression>,
-    pub statements: Vec<Statement>,
-    pub sampler: Option<Handle<Expression>>,
+#[derive(Debug, Clone)]
+pub struct VariableReference {
+    pub expr: Handle<Expression>,
+    pub load: Option<Handle<Expression>>,
 }
 
-impl ExpressionRule {
-    pub fn from_expression(expression: Handle<Expression>) -> ExpressionRule {
-        ExpressionRule {
-            expression,
-            statements: vec![],
-            sampler: None,
-        }
-    }
+#[derive(Debug, Clone)]
+pub struct Expr {
+    pub kind: ExprKind,
+    pub meta: TokenMetadata,
+}
+
+#[derive(Debug, Clone)]
+pub enum ExprKind {
+    Access {
+        base: Box<Expr>,
```

Comment:
ooh, that's a bit heavy-handed. Many boxes!
Perhaps, GLSL frontend doesn't need to be as fast as the others, but everywhere else we stay away from such boxes.
Perhaps, we can instead just use the arena here? I.e. that would be `Handle<Expr>` instead of `Box<Expr>`. Although, I'm not sure what purpose would `enum ExprKind` have in this case - it would look very similar to `enum Expression`.

Reply:
I'm planning to replace the boxes with an arena just haven't gotten around to it (also I wonder if it would be possible to make the lowering in a loop trough the arena instead of recursive).

The `ExprKind` exists to handle stores, LHS expressions and in the future postfix and prefix Inc/Dec.

Reply:
This part now uses a `Arena` there's still an annoying clone in the lowering part (It's possible to remove but it would require some organizational changes that I think should be done in another PR) but it's a shallow clone and should much better than the previous implementation.